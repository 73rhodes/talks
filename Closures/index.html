<!doctype html>
<html lang="en">

    <head>
        <meta charset="utf-8">

        <title>Closures for 6 yr olds</title>

        <meta name="description" content="A simple introduction to JavaScript closures by example.">
        <meta name="author" content="Darren DeRidder">

        <meta name="apple-mobile-web-app-capable" content="yes" />
        <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

        <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

        <link rel="stylesheet" href="css/reveal.min.css">
        <link rel="stylesheet" href="css/theme/solarized.css" id="theme">

        <!-- For syntax highlighting -->
        <link rel="stylesheet" href="lib/css/zenburn.css">

        <!-- If the query includes 'print-pdf', use the PDF print sheet -->
        <script>
            document.write( '<link rel="stylesheet" href="css/print/' + ( window.location.search.match( /print-pdf/gi ) ? 'pdf' : 'paper' ) + '.css" type="text/css" media="print">' );
        </script>

        <!--[if lt IE 9]>
        <script src="lib/js/html5shiv.js"></script>
        <![endif]-->
    </head>

    <body>

        <div class="reveal">

            <!-- Any section element inside of this container is displayed as a slide -->
            <div class="slides">
                <section>
                    <h2>JavaScript Closures</h2>
                    <h3>for 6 year olds</h3>
                    <p>
                    <small>
                      <a href="http://51elliot.blogspot.com">
                        Darren DeRidder
                      </a>
                      / 
                      <a href="http://twitter.com/73rhodes">
                        @73rhodes
                      </a>
                    </small>
                    </p>
                </section>

                <section>
                        <h2>What's a closure?</h2>
                        <blockqoute cite="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Closures">
                        "Closures are functions that refer to independent
                        (free) variables." [MDN]
                        </blockquote>
                 </section>

                 <section>
                       <h3>Why Closures?</h3>
                       <ul>
                         <li>Common in JavaScript.</li>
                         <li>Node.JS callbacks.</li>
                         <li>Some practical uses.</li>
                         <li>Some unexpected problems.</li>
                       </ul>
                 </section>

		<section data-background="feynman.jpg">
                  <blockquote style="color: white; background: rgba(64,64,64,0.6);">
		  "If you can't explain it to a six year old, you don't really understand it." - Richard Feynman
                  </blockquote>
		</section>

                <section>
                    <h2>The Simplest Closure</h2>
                    <pre><code data-trim contenteditable>
&nbsp;
    var x = 1;
    function foo() {
      return x;
    }
&nbsp;
                    </code></pre>
                </section>

        <section data-background="travel.png">
		<h3>The Traveller</h3>
                <blockquote style="color: black; background: rgba(230,225,220,0.9);">
		<p>
		A man goes on a long trip. Before leaving, he takes a snapshot
                of his house. 
                <p>
                While he's away, the house is replaced with a 
                condo. The man is unaware and continues to show the photo and 
                describe his house to the people he meets.
		</blockquote>
        </section>

                <section>
                    <h3>The Traveller code example</h3>
                    <pre><code data-trim contenteditable>
&nbsp;
    var color = 'white';

    function leaveHome () {
       // a variable in leaveHome's scope
       var description = "My house is " + color;
       // returns another function, creating a closure over description
       return function() {
          console.log(description);
       }
    }

    var travellingMan = leaveHome();
    travellingMan(); // "My house is white"
    color = 'red';   // They painted the house
    travellingMan(); // "My house is white"
&nbsp;
                    </code></pre>
                </section>

                <section>
                    <h3>Things to Notice</h3>
                    <ol>
		      <li>In Javascript, functions are objects, too.</li>
                      <li>Functions can create and return other functions.</li>
                      <li>Functions can be anonymous.</li>
		      <li>The returned function "closes over" the variables in
                          the scope where it was created, forming a closure.</li>
                    </ol>
                </section>
                <section>
                   <h3>More examples</h3>
                    <ul>
                      <li>An "inner function" is a closure.</li>
                      <li>Functions that get returned are closures.</li>
                      <li>Functions that get passed (ie. callbacks) are closures.</li>
                      <li>Any function that accesses variables outside it's immediate lexical scope is a closure.</li>
                    </ul>
               </section>

                <section>
                    <h3>Related Methods</h3>
                     <ul>
                     <li>Function.call</li>
                     <li>Function.apply</li>
                     <li>Function.bind</li>
                     <ul>
                </section>

                <section>
                <h3>Thanks!</h3>
                <a href="http://darrenderidder.github.io/talks/Closures">
                http://darrenderidder.github.io/talks/Closures
                </a>
                </section>

            </div>

        </div>

        <script src="lib/js/head.min.js"></script>
        <script src="js/reveal.min.js"></script>

        <script>

            // Full list of configuration options available here:
            // https://github.com/hakimel/reveal.js#configuration
            Reveal.initialize({
                margin: 0,
                controls: true,
                progress: true,
                history: true,
                center: true,

                theme: Reveal.getQueryHash().theme, // available themes are in /css/theme
                transition: Reveal.getQueryHash().transition || 'default', // default/cube/page/concave/zoom/linear/fade/none

                // Optional libraries used to extend on reveal.js
                dependencies: [
                    { src: 'lib/js/classList.js', condition: function() { return !document.body.classList; } },
                    { src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
                    { src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
                    { src: 'plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
                    { src: 'plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
                    { src: 'plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
                ]
            });

        </script>

    </body>
</html>
