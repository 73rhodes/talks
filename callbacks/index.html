<!doctype html>
<html lang="en">

  <head>
    <meta charset="utf-8">

    <title>I &lt;3 Callbacks</title>

    <meta name="description" content="Taming the Pyramid of Doom the Simple Way">
    <meta name="author" content="Darren DeRidder">

    <meta name="apple-mobile-web-app-capable" content="yes" />
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <link rel="stylesheet" href="css/reveal.min.css">
    <link rel="stylesheet" href="css/theme/solarized.css" id="theme">

    <!-- For syntax highlighting -->
    <link rel="stylesheet" href="lib/css/zenburn.css">

    <!-- If the query includes 'print-pdf', include the PDF print sheet -->
    <script>
      if( window.location.search.match( /print-pdf/gi ) ) {
        var link = document.createElement( 'link' );
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = 'css/print/pdf.css';
        document.getElementsByTagName( 'head' )[0].appendChild( link );
      }
    </script>

    <!--[if lt IE 9]>
    <script src="lib/js/html5shiv.js"></script>
    <![endif]-->
  </head>

  <body>

    <div class="reveal">

      <!-- Any section element inside of this container is displayed as a slide -->
      <div class="slides">
        <section>
          <h1>I &lt;3 Callbacks</h1>
          <h3>Taming the Pyramid of Doom</h3>
          (the simple way)
        </section>

        <section>
          <h2>About Me</h2>
          <p>
            <small><a href="http://about.me/drderidder">Darren DeRidder</a> / <a href="http://twitter.com/73rhodes">@73rhodes</a></small><br/>
            Alcatel-Lucent - Motive Mobile Security<br>
            Real-time, AAA, Telecom, Network &amp; Mobile Security<br>
            Mobile World Congress<br/>
            Blackhat Security Conference<br>
            Ottawa JavaScript<br>
          </p>

          <aside class="notes">
            Thanks to:<br/>
            Thomas, David, Dan, Simon M, Simon K, Owain, Nick.
            Mark, Jillian, Randy, Fabien.
            Toushay, Teldio, CanvasPop, Magmic Games, IBM, MXI, Shopify, Adobe
          </aside>
        </section>

        <section>

          <section>
            <h2>Synchronous</h2>
            <img src="./linearity.png"></img>
          </section>

          <section>
            <h3>Synchronous Programming</h3>
            <ul>
              <li>Linear</li>
              <li>Easy to understand</li>
              <li>Functions return values</li>
              <li>Code reads like a story</li>
            </ul>
          </section>

          <section>
            <h3>Synchronous Code</h3>
              <pre><code data-trim contenteditable>
function add (a, b) {
  return a + b;
}

var c = add (9, 33);

myElement.innerHTML = c;
              </pre></code>
          </section>

        </section>

        <section>

          <section>
            <h2>Asynchronous</h2>
            <img src="./nonlinearity.png"></img>
          </section>

          <section>
            <h3>Asynchronous Programming</h3>
            <ul>
              <li>non-linear</li>
              <li>harder to reason about</li>
              <li>choose-your-own-adventure</li>
              <li>callbacks</li>
            </ul>
          </section>

          <section>
            <h3>Asynchronous APIs</h3>
            <ul>
              <li>AJAX</li>
              <li>HTML5 Geolocation</li>
              <li>HTML5 Web Workers</li>
              <li>IndexedDB</li>
              <li>Node.JS</li>
            </ul>
          </section>

          <section>
            <h3>Asynchronous Code</h3>
            <pre><code data-trim contenteditable>
$.ajax({
  type: 'POST',
  url: "path/to/api",
  data: "banana=yellow",
  success: function (data) {
    alert("Success: " + data);
  }
});
            </code></pre>
          </section>

        </section>

        <section>
          <section>
            <h2>Continuation Passing Style</h2>
            <pre style="background:#EFEFEF;">

 do something (
     with these parameters,
     afterwards call this function with the result
 )
            </pre>
            <pre><code data-trim contenteditable>
fs.readFile('todo.txt', function(err, data) {
  if (err) {
    console.log('Error reading file: ' + err)
  } else {
    console.log(data);
  }
});
            </code></pre>
            <p class="fragment">Leads to...</p>
          </section>

          <section>
            <h3>Callback Hell</h3>
              <pre><code data-trim contenteditable>
fs.readdir(source, function(err, files) {
  if (err) {
    console.log('Error finding files: ' + err)
  } else {
    files.forEach(function(filename, fileIndex) {
      console.log(filename)
      gm(source + filename).size(function(err, values) {
        if (err) {
          console.log('Error identifying file size: ' + err)
        } else {
          console.log(filename + ' : ' + values)
          aspect = (values.width / values.height)
          widths.forEach(function(width, widthIndex) {
            height = Math.round(width / aspect)
            console.log('resizing ' + filename + 'to ' + height + 'x' + height)
            this.resize(width, height).write(destination + 'w' + width + '_' + filename, function(err) {
              if (err) console.log('Error writing file: ' + err)
            })
          }.bind(this))
        }
      })
    })
  }
})
            </code></pre>
            <p class="fragment">Leads to...</p>
          </section>
 
          <section>
            <h3>Control Flow Libraries</h3>
            <p class="fragment">lots and lots of them</p>
            <p class="fragment">
            yuemeng $N async-mini async.js async backbone-callbacks begin condvar 
            chainsaw channels cloudd ctrl deferred deferred-queue each eventproxy
            finish first flow-js flowless funk futures galaxy gate gene groupie
            ice-stream Ignite jam JobManager jscex-jquery LAEH2 locker make-flow
            miniqueue narrow nestableflow neuron node-async node-block node-chain
            node-continuables node-cron node-crontab node-fnqueue node-inflow
            node-parallel node-promise node.flow noflo nThen nue observer pausable
            pipette poolr promise q QBox Rubberduck statechart SCION serialize
            seq Signals simple-schedule Slide sode.js Step stepc streamify 
            suspend taskman timequeue TwoStep when zo.js y-yield ... 
            </p>
          </section>

          <section>
            asyncblock asyncflow fiberize fibers-promise fibers sync Cinch
            Wind.js qjs sjs streamline.js syncasync.js tamejs promises 
            wsl-promises mock-promises promises-promises ql naive-promesse
            spai bluebid pacta avec yapa th tillthen microromise await qall
            noq sinon-promise pangako deferredjs este-dispather ahead jqx
            yapawapa fate callbackify then-all troth augur neder avow prfun
            qdomin clues minq then-node zee pooldis assure qplux z-core
            bluecache resolved dom-futures risposta memoblock concurrent
            fantasy-async heya-async pgpromise q-step orchard legendary
            covenant qx conterfeit faithful qpatch noob_pact fsp 
            stepthrough diva milestonesjs qtree psqueue flourine qsub
            majic async-done underpromise ninvoke then-all-settled-in 
            then-all-in execution-pool prim kew d.js ncall q-lazy lag

            </p>
            <p>
              and 1,847 more...
            </p>
          </section>
        </section>

        <section>
          <section>
            <h2>Promises?</h2>
            <p>Removed from Node core in 2010 by Ryan</p>
            <p class="fragment">
            Suggesting that promises should go back in core
            <img src="http://media.tumblr.com/3a135ef80fbb238fd30154ccd38e604d/tumblr_mjh2luxMTr1s54fggo1_500.gif"></img>
             <small>
             <a href="http://nodejsreactions.tumblr.com/post/65344663524">http://nodejsreactions.tumblr.com/post/65344663524 </a>
             </small>
            </p>
          </section>

          <section>
            <img src="./tjonpromises.png"></img>
            <p class="fragment">
            ...plus Trever Norris, Isaac Schlueter, Fedor Idutny, Mikeal Rogers, Ryan Dahl,
            Rodd Vagg...
            (the whole node core dev team)
            </p>
          </section>

          <section>
            <h3>Node core philosophy</h3>
            <ul>
              <li>decisions that lead to maximum compatibility</li>
              <li>keep core small and light</li>
              <li>high level flow-control abstractions in user libs</li>
            </ul>
          </section>

          <section>
            <h3>some thoughts on promises</h3>
            <ul>
              <li>
                <a href="http://andrewkelley.me/post/js-callback-organization.html">Callbacks are really okay</a>
              </li>
              <li>
                <a href="http://sealedabstract.com/code/broken-promises/">Promises have unexpected costs</a>
              </li>
              <li>
                <a href="http://r.va.gg/presentations/webdirections-code-2014/#20">Callbacks are the fundamental unit of async in JS.</a>
              </li>
              <li>
                <a href="http://eamodeorubio.github.io/tamingasync/#/23">Promises don't work for many use cases</a>
              </li>
              <li>
                <a href="http://www.futurealoof.com/posts/broken-promises.html">Flow control not defined in core by design</a>
              </li>
              <li>
                <a href="http://blog.trevnorris.com/2014/02/nodejs-es6-and-me.html">Integrating c++ / libuv only possible with callbacks</a>
              </li>
            </ul>
          </section>
 
        </section>

        <section>
          <h3>Simple Rules for Escaping Callback Hell</h3>
          <ol>
            <li>Use named functions for callbacks</li>
            <li>Use nesting only to enclose variable scope</li>
            <li>Return when invoking the callback</li>
          </ol>
        </section>

        <section>
          <h3>The Pyramid of Doom</h3>
          <img style="height: 600px;" src="./callbacks1.png"></img>
        </section>

        <section>
          <section>
            <h3>Named Callbacks</h3>
            <ul>
              <li>Avoid non-trivial anonymous functions</li>
              <li>Put function declarations after code that uses them</li>
            </ul>
            <pre><code data-trim contenteditable>

somethingAsync(params, namedCallback);

function namedCallback(err, result) { 
  // handle error, result
}

            </code></pre>
          </section>

          <section>
            <h3>Named Callbacks</h3>
            <img style="height: 600px;" src="./callbacks2.png"></img>
          </section>

        </section>

        <section>
          <section>
            <h3>Nest for Scope</h3>
             Nested functions share variables in the enclosing scope
            <pre><code data-trim contenteditable>
function foo() {
  var x = 1;            // in foo's scope

  function bar() {
    var y = 2;          // in bar's scope
    console.log(x + y); // 3
  }

  function fizz() {
    var z = 3;          // in fizz's scope
    console.log(x + z); // 4
  }

  console.log(y); // undefined
}
            </code></pre>
          </section>

          <section>
            <h3>Nest for Scope</h3>
            <img style="height: 600px;" src="./callbacks3.png"></img>
          </section>

        </section>

        <section>
          <section>
            <h3>Return when invoking callbacks</h3>
            <pre><code data-trim contenteditable>
function myAsyncFunction (params, callback) {

  // You do something asynch...
  
  // You handle the result
  function handleResult(err, result) {
    if (err) {
      return callback(err);
    }
    // do something with data
    return callback(null, data);
  }

}

            </code></pre>
          </section>

          <section>
            <h3>Return when invoking callbacks</h3>
            <img style="height: 600px;" src="./callbacks4.png"></img>
          </section>
        </section>

        <section>
          <h3>Further Reading</h3>
          <ul>
            <li><a href="callbackhell.com">callbackhell.com</a></li>
            <li><a href="http://andrewkelley.me/post/js-callback-organization.html">JavaScript Callbacks Are Pretty Okay</a></li>
            <li><a href="http://51elliot.blogspot.ca/2014/11/three-simple-rules-for-escaping.html">Three Simple Rules for Escaping Callback Hell</a></li>
          </ul>
        </section>

        <section>
          <h2>Thanks!</h2>
          http://darrenderidder.github.io/talks/callbacks
        </section>
      </div>

    </div>

    <script src="lib/js/head.min.js"></script>
    <script src="js/reveal.min.js"></script>

    <script>

      // Full list of configuration options available here:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        controls: true,
        progress: true,
        history: true,
        center: true,

        theme: Reveal.getQueryHash().theme, // available themes are in /css/theme
        transition: Reveal.getQueryHash().transition || 'default', // default/cube/page/concave/zoom/linear/fade/none

        // Parallax scrolling
        // parallaxBackgroundImage: 'https://s3.amazonaws.com/hakim-static/reveal-js/reveal-parallax-1.jpg',
        // parallaxBackgroundSize: '2100px 900px',

        // Optional libraries used to extend on reveal.js
        dependencies: [
          { src: 'lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
          { src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
          { src: 'plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
          { src: 'plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
          { src: 'plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
        ]
      });

    </script>

  </body>
</html>
