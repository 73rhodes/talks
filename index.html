<!DOCTYPE html>
<html lang="en">
<head>

  <!-- Basic Page Needs
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
  <meta charset="utf-8">
  <title>Talks by @73rhodes</title>
  <meta name="description" content="">
  <meta name="author" content="">

  <!-- Mobile Specific Metas
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <!-- FONT
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
  <link href="//fonts.googleapis.com/css?family=Raleway:400,300,600" rel="stylesheet" type="text/css">

  <!-- CSS
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
  <link rel="stylesheet" href="css/normalize.css">
  <link rel="stylesheet" href="css/skeleton.css">

  <!-- Favicon
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
  <!--
  <link rel="icon" type="image/png" href="images/favicon.png">
  -->

</head>
<body>

  <!-- Primary Page Layout
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
  <div class="container">
    <div class="row">
      <div class="one-half column" style="margin-top: 25%">
        <h4>Talks by @73rhodes</h4>
        <ul>
          <li>
            <a href="http://73rhodes.github.io/talks/MachineLearning/">Machine Learning with Node.JS</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/ProcessingJS/">Intro to Processing.JS</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/ModulePatterns/">Node.JS Module Patterns with Simple Examples</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/Closures/">JavaScript Closures for 6 Year Olds</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/callbacks/">I &lt;3 Callbacks - Taming the Pyramid of Doom</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/functional-intro/">Intro to Functional Programming in JS</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/Golden/">Secret Canons of Medieval Incunabula</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/getting-started-with-js/">Getting Started with JavaScript</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/multicore/">Scaling CPU-intensive Node.JS Apps on Multi-core Systems</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/FlowIntro/README.md">Flow Intro</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/jswidgets/README.md">JS Widgets</a>
          </li>
          <li>
            <a href="http://73rhodes.github.io/talks/jsdoctypes">Types with JSDoc</a>
          </li>
        </ul>
      </div>
    </div>
  </div>

<!-- End Document
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
</body>
</html>
