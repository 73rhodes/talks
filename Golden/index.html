<!doctype html>
<html lang="en">

  <head>
    <meta charset="utf-8">

    <title>Canons of Medieval Incunabula for the Modern Web</title>

    <meta name="description" content="Explorations of the Golden Mean and Canons of Layout">
    <meta name="author" content="Darren DeRIdder">

    <meta name="apple-mobile-web-app-capable" content="yes" />
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">

    <link rel="stylesheet" href="css/reveal.css">
    <link rel="stylesheet" href="css/theme/black.css" id="theme">

    <!-- Code syntax highlighting -->
    <link rel="stylesheet" href="lib/css/zenburn.css">

    <!-- Printing and PDF exports -->
    <script>
      var link = document.createElement( 'link' );
      link.rel = 'stylesheet';
      link.type = 'text/css';
      link.href = window.location.search.match( /print-pdf/gi ) ? 'css/print/pdf.css' : 'css/print/paper.css';
      document.getElementsByTagName( 'head' )[0].appendChild( link );
    </script>

    <style>
    section.phi * {
      /*font-size: 1.5em;*/
      font-family: "Helvetica Neue", sans-serif;
      font-weight: 100;
      text-shadow: 2px 2px 6px #000000;
    }
    </style>

    <!--[if lt IE 9]>
    <script src="lib/js/html5shiv.js"></script>
    <![endif]-->
  </head>

  <body>

    <div class="reveal">

      <!-- Any section element inside of this container is displayed as a slide -->
      <div class="slides">

        <section class="phi" data-background="images/typgn.jpg">
          <h3>Secret Canons</h3>
        </section>

        <section class="phi">
          <h2>Secret Canons</h2>
          <p>of Medieval Incunabula</p>
          <p>The Golden Mean, Modernist Layout and the Web</p>
        </section>

        <!-- Example of nested vertical slides -->
        <section>
          <section class="phi">
            <p>Hi, I'm Darren</p>
            <p>@73rhodes</p>
          </section>
          <section class="phi">
            <p>Bridgewater. Kindsight. Alcatel-Lucent. Shopify.</p>
            <p>Telecom. AAA. Security. Mobile.</p>
            <p>Mobile World Congress. Blackhat.</p>
          </section>
        </section>

        <section>
          <section class="phi">
            <h3>Layout</h3>
            <p>Canonical or Accidental</p>
          </section>
          <section class="phi">
            <img src="images/screencap1.jpg"/>
            <p>grids</p>
          </section>
          <section class="phi">
            <img src="images/screencap3.jpg"/>
            <p>templates</p>
          </section>
          <section class="phi">
            <img src="images/bad-website2.jpg"/>
            <p>accidental?</p>
          </section>
          <section class="phi">
            <img src="images/bad-website1.jpg"/>
            <p>wtf?</p>
          </section>

        </section>

        <section>
          <section class="phi" data-background="images/incunabula.jpg">
            <h1>Incunabula</h1>
          </section>

          <section class="phi">
              <img src="images/incunabula139.jpg"></img>
              <p>"Phisicorum", HMD Collection, c. 1485</p>
          </section>

          <section class="phi">
            <img src="images/incunabula-3.jpg"></img>
            <p>University of Glasgow Collection, c. 1465</p>
          </section>

          <section class="phi">
            <img src="images/Gutenberg_Bible.jpg"></img>
            <p>Gutenberg Bible, Johannes Gutenberg, c. 1455</p>
          </section>

          <section class="phi">
            <img src="images/germ-1496-verso.jpg"></img>
            <p>unknown, 15th Century Germany
          </section>

        </section>

        <section>
          <section class="phi">
            <div>
              <img style="float: left;" src="images/tschichold_schreibt.jpg"/>
              <h2>Jan Tschichold</h2>
              <p>1902 - 1974</p>
              <p>
                The Form of the Book
              </p>
            </div>
          </section>
          <section class="phi">
            <img src="images/Bauhaus.jpg"/>
            <p>
              The Bauhaus School of Design, Dessau
            </p>
          </section>
          <section class="phi">
            <h2>The Golden Canon</h2>
            <img src="images/golden_canon.jpg"/>
          </section>

        </section>

        <section>
          <section class="phi">
            <img style="float: left;"src="images/rosarivo1.jpg"/>
            <h2>Raúl Rosarivo</h2>
            <p>1903 - 1966</p>
            <p>
              Divina proporción tipográfica, Argentina, 1953
            </p>
          </section>
          <section class="phi">
            <h2>Rosarivo's Canon</h2>
            <img src="images/rosarivo_canon.jpg"/>
          </section>
        </section>

        <section>
          <section class="phi">
            <h2>J. A. Van de Graaf</h2>
            <p>c. 1900</p>
            <p>"Nieuwe berekening voor de vormgeving" in Tete, Amsterdam, 1946</p>
          </section>
          <section class="phi">
            <h2>The Secret Canon</h2>
            <img src="images/vandegraaf_canon.jpg"/>
          </section>
        </section>

        <section>
          <section class="phi">
            <img style="float: left;" src="images/Villard_de_Honnecourt.jpg"/>
            <h2>Villard de Honnecourt</h2>
            <p>Folio, Picardy, 33 pages, c. 1230</p>
          </section>
          <section class="phi">
            <img style="float: left;" src="images/Villard_de_Honnecourt_29.jpg"/>
            <h3>Villard Folio</h3>
            <p>p. 29</p>
            <a href="http://commons.wikimedia.org/wiki/Villard_de_Honnecourt">View on Wikimedia Commons</a>
          </section>
          <section class="phi">
            <img style="float: left;" src="images/villard_diagram.jpg"/>
            <h3>The "Villard Diagram"</h3>
            <ul>
              <li>Attribution by Tschichold</li>
              <li>Divide into N equal parts</li>
              <li>Provenance unclear</li>
            </ul>
          </section>
        </section>

        <section>
          <section class="phi">
            <h3>The Golden Ratio</h3>
            <p>A / B = (A + B) / A</p>
            <span style="font-family: cursive;">φ = 1.61803398875</span>
          </section>
          <section class="phi">
            <h3>The Golden Rectangle</h3>
            <img src="images/gold_rec.gif"/>
          </section>
          <section class="phi">
            <img src="images/GoldenSpiral.gif"/>
          </section>
          <section class="phi">
            <img src="images/parthenon-golden-ratio.jpg"/>
          </section>
          <section class="phi">
            <img style="float: left;" src="images/davinci_vitruvian_man.jpg"/>
            <h3>Leonardo da Vinci</h3>
            <p>Vitruvian Man, c. 1490</p>
          </section>
          <section class="phi">
            <img style="float: left; margin-right: 25px;" src="images/ferrara_vitruvian_man.jpg"/>
            <h3>Giacomo Andrea Da Ferrara</h3>
            <p>Vitruvian Man, c.1490</p>
          </section>
          <section class="phi">
            <h3>Marcus Vitruvius Pollio</h3>
            <p><i>Di Architura</i></p>
            <p>c.80 BC - c. 15 BC</p>
          </section>
          <section class="phi">
            <img style="float: left; margin-right: 25px;" src="images/Euclid.jpg"/>
            <h3>Euclid</h3>
            <p><i>Elements</i>, c. 300 BC</p>
            <p>"A straight line is said to have been cut in extreme and mean ratio when, as the whole line is to the greater segment, so is the greater to the lesser."</p>
          </section>
          <section class="phi">
            <h3>Plato</h3>
            <p><i>Timaeus</i>, c.360 BC</p>
            <p>"For whenever in any three numbers, whether cube or square, there is a mean, which is to the last term what the first term is to it; and again, when the mean is to the first term as the last term is to the mean—then the mean becoming first and last, and the first and last both becoming means, they will all of them of necessity come to be the same, and having become the same with one another will be all one"</p>
          </section>
        </section>

        <section>
          <section class="phi">
            <h3>The Modern Web</h3>
          </section>
          <section class="phi">
            <p>
              a scrolling medium
            </p>
          </section>
          <section class="phi">
            <p>
              Arbitrary text areas -
            </p>
            <p>canons for book layout won't work</p>
          </section>
          <section class="phi">
            <p>
              CSS grid systems
              <ul>
                <li>Bootstrap</li>
                <li>Foundation</li>
                <li>960, etc.</li>
              </ul>
            </p>
          </section>
          <section class="phi">
            <p>
              <a href="https://73rhodes.github.io/phi">Phi.JS</a>
              - Secret Canons on the Web
            </p>
            <ul>
              <li>Van de Graaf's 'Secret Canon'</li>
              <li>Tschichold (new, based on Tschichold's Golden Canon)</li>
              <li>Optical Centering</li>
              <li>new: "Neo-Tschichold"</li>
            </ul>
          </section>
        </section>


        <section class="phi">
          <h1>finis</h1>
        </section>

      </div>

    </div>

    <script src="lib/js/head.min.js"></script>
    <script src="js/reveal.js"></script>

    <script>

      // Full list of configuration options available at:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        controls: true,
        progress: true,
        history: true,
        center: true,

        transition: 'zoom', // none/fade/slide/convex/concave/zoom

        // Optional reveal.js plugins
        dependencies: [
          { src: 'lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
          { src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
          { src: 'plugin/highlight/highlight.js', async: true, condition: function() { return !!document.querySelector( 'pre code' ); }, callback: function() { hljs.initHighlightingOnLoad(); } },
          { src: 'plugin/zoom-js/zoom.js', async: true },
          { src: 'plugin/notes/notes.js', async: true }
        ]
      });

    </script>

  </body>
</html>
